### PR TITLE
chore: support multi-org group sync with runtime configuration

### LINKED ISSUE #14202: Multi-Organization Group Sync compatibility upgrade
Migrate feature away from using the default organization.

At present [`oidcGroups`](https://github.com/coder/coder/blob/8c0565177e5bc5611130df54cf857f9e6a468027/coderd/userauth.go#L1083-L1083) returns a flat `[]string` of groups to assign to the user. This flat list is applied to the [default organization](https://github.com/coder/coder/blob/8c0565177e5bc5611130df54cf857f9e6a468027/coderd/userauth.go#L1547-L1549). 

# Organization Membership

Group sync could, and might be extended to membership sync. Meaning we might want to use the "everyone" group to assign users into an org.

So the default config would look like (json structure subject to change):

```
# assign the oidc group to the 'everyone' group in the default org. So
# these users are placed into the "default" org as a member. 
CODER_OIDC_GROUP_MAPPING='{"myOIDCGroupID": {"default":"everyone"}}'
```

This means if the user has no `myOIDCGroupID` group, **they would be removed from the default org**. Using groups to assign org membership would be great if we can handle all the edge cases and defaults. 

- Should there be something like `CODER_OIDC_GROUP_ALWAYS='{"default":"everyone"}'` to make sure everyone is in the default org?
- Easy to make orgless users.

Some design principles to consider. Organizations are intended to be "self serve" from the perspective of an organization admin. It reasons then that a site wide admin might control which users are placed into an org, but further group membership could be decided by the admins of said org.

This could potentially split the configuration into 2 places. 

# Deployment configuration to upgrade

## Auto-create

The [`CODER_OIDC_GROUP_AUTO_CREATE`](https://github.com/coder/coder/blob/8c0565177e5bc5611130df54cf857f9e6a468027/codersdk/deployment.go#L1558-L1567) poses a challenge in multi-org. As it is no longer obvious which organization the new group should be created in.

## Static group mapping

The static mapping [`CODER_OIDC_GROUP_MAPPING`](https://github.com/coder/coder/blob/8c0565177e5bc5611130df54cf857f9e6a468027/codersdk/deployment.go#L1548-L1557) needs to accommodate an organization context. This map
 

JSON is currently `map[<OIDC_group_name>]<coder_group_name>`, so an example of the **current** configuration is:

```
# Flat assignment assuming default org. We need to inject an `<ord_id>` into this
CODER_OIDC_GROUP_MAPPING='{"myOIDCGroupID": "myCoderGroupName"}'
```

We need a json structure that includes organization context for the group. For single org deployments, maybe the old config should still work?

### PR DESCRIPTION
**Pull request by Emyrk**
Closes https://github.com/coder/coder/issues/14202

# What is this

Group sync implemented to support `multi-organization` runtime configuration (CRUD does not yet exist).

All previous logic existed in `oidcGroups` and `oauthLogin`. I disliked this approach, as the components were completely untestable. So tests **had to be** integration tests.

The goal of this package is to completely test the sync logic independent of the oidc or broader coderd code.

# What is does

Enables group sync to work in multiple organizations, not just the default. There is no way to configure other organizations in this PR, the API and CRUD is not done yet.

Unit tests show how these values can be set.

# Testing notes

This is a large refactor to a section of code that had very little testing of it's parts. It did however have good integration style tests for group sync **in the default org**. In the package tests, multi-org sync is tested, but only with `dbmem` as it inserts a lot of data. When CRUD api is added, more tests can be added similar to those below for multi-org.

- https://github.com/coder/coder/blob/stevenmasley%2Forg_group_sync/enterprise/coderd/userauth_test.go#L710-L710
- https://github.com/coder/coder/blob/stevenmasley%2Forg_group_sync/enterprise/coderd/userauth_test.go#L392-L392
- https://github.com/coder/coder/blob/stevenmasley%2Forg_group_sync/coderd/userauth_test.go#L866-L866

### CODE REVIEW

File: agent/proto/agent.proto [hunk 1]
```diff
@@ -41,6 +41,7 @@ message WorkspaceApp {
 		UNHEALTHY = 4;
 	}
 	Health health = 12;
+	bool hidden = 13;
```
Comment (f0ssel) [inline agent/proto/agent.proto:44]: Is this suppose to be part of this change set? Seems unrelated?

**Review by f0ssel — COMMENTED** [review]
Overall looks good, glad the runtime config only impacts this feature. Unfortunate about the sql tests being disabled, maybe we can add some sort of simple cache or debounce for just tests? Could be a followup. 

Also it seems like all the frontend changes should be in a separate PR? Or maybe the branch just needs rebasing?

**Comment by Emyrk** [comment]
@f0ssel yea, I rebased it correctly, but forgot to change the base branch to `main` :facepalm:. Should be correct now :+1:

**Review by f0ssel — APPROVED** [review]

File: coderd/coderdtest/uuids_test.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		_ = tc
+		// Do the test with CreateUsers as the setup, and the expected IDs
+		// will match.
+	}
```
Comment (coadler) [inline coderd/coderdtest/uuids_test.go:32]: This doesn't seem to be running the test cases?

File: coderd/database/queries/groups.sql [hunk 3]
```diff
@@ -52,6 +52,10 @@ WHERE
 						)
 				ELSE true
 		END
+		AND CASE WHEN array_length(@group_names :: text[], 1) > 0  THEN
+					 groups.name = ANY(@group_names)
+			ELSE true
+		END
```
Comment (coadler) [inline coderd/database/queries/groups.sql:58]:
```suggestion
		AND CASE WHEN array_length(@group_names :: text[], 1) > 0  THEN
				groups.name = ANY(@group_names)
			ELSE true
		END
```
nit; indentation is a little weird

**Review by coadler — APPROVED** [review]
still trying to catch up to speed but all lgtm!

File: coderd/coderdtest/uuids_test.go (hunk 2 above)
Comment (Emyrk) [reply coderd/coderdtest/uuids_test.go:32]: :facepalm: 

I had this in another local branch at first. I'll bring in the correct code, good catch.

Comment (Emyrk) [reply coderd/coderdtest/uuids_test.go:32]: This is a [TestableExample](https://go.dev/blog/examples). Looks like I kinda gave up here. I'm going to just remove. I have example usage in the idp package now.

File: coderd/database/dbmem/dbmem.go [hunk 4]
```diff
@@ -682,6 +682,17 @@ func (q *FakeQuerier) getWorkspaceResourcesByJobIDNoLock(_ context.Context, jobI
...
+func (q *FakeQuerier) getGroupByNameNoLock(arg database.NameOrganizationPair) (database.Group, error) {
+	for _, group := range q.groups {
+		if group.OrganizationID == arg.OrganizationID &&
+			group.Name == arg.Name {
+			return group, nil
+		}
+	}
+
+	return database.Group{}, sql.ErrNoRows
+}
```
[the remark this answers is not in the document]
Comment (Emyrk) [reply coderd/database/dbmem/dbmem.go:694]: Oh, I originally was using it in another query. Wanted to reuse the same code. Now it's not necessary to extract to it's own function :man_shrugging:. I'll put it back.
